ibis/pyspark/compiler.py

```diff
@@ -0,0 +1,514 @@
+
+    return src_table[col_names_in_selection_order]
+
+
+@compiles(ops.TableColumn)
+def compile_column(t, expr, scope, **kwargs):
+    op = expr.op()
+    table = compile_with_scope(t, op.table, scope)
+    return table[op.name]
+
+
+@compiles(ops.SelfReference)
+def compile_self_reference(t, expr, scope, **kwargs):
+    op = expr.op()
+    return t.translate(op.table, scope)
+
+
+@compiles(ops.Equals)
+def compile_equals(t, expr, scope, **kwargs):
+    op = expr.op()
+    return t.translate(op.left, scope) == t.translate(op.right, scope)
+
+
+@compiles(ops.Greater)
+def compile_greater(t, expr, scope, **kwargs):
+    op = expr.op()
+    return t.translate(op.left, scope) > t.translate(op.right, scope)
+
+
+@compiles(ops.GreaterEqual)
+def compile_greater_equal(t, expr, scope, **kwargs):
+    op = expr.op()
+    return t.translate(op.left, scope) >= t.translate(op.right, scope)
+
+
+@compiles(ops.Multiply)
+def compile_multiply(t, expr, scope, **kwargs):
+    op = expr.op()
+    return t.translate(op.left, scope) * t.translate(op.right, scope)
+
+
+@compiles(ops.Subtract)
+def compile_subtract(t, expr, scope, **kwargs):
+    op = expr.op()
+    return t.translate(op.left, scope) - t.translate(op.right, scope)
+
+
+@compiles(ops.Literal)
+def compile_literal(t, expr, scope, raw=False, **kwargs):
+    """ If raw is True, don't wrap the result with F.lit()
+    """
+    value = expr.op().value
+
+    if raw:
+        return value
+
+    if isinstance(value, collections.abc.Set):
+        # Don't wrap set with F.lit
+        if isinstance(value, frozenset):
+            # Spark doens't like frozenset
+            return set(value)
+        else:
+            return value
+    else:
+        return F.lit(expr.op().value)
+
+
+@compiles(ops.Aggregation)
+def compile_aggregation(t, expr, scope, **kwargs):
+    op = expr.op()
+
+    src_table = t.translate(op.table, scope)
+
+    if op.by:
+        context = AggregationContext.GROUP
+        aggs = [t.translate(m, scope, context=context)
+                for m in op.metrics]
+        bys = [t.translate(b, scope) for b in op.by]
+        return src_table.groupby(*bys).agg(*aggs)
+    else:
+        context = AggregationContext.ENTIRE
+        aggs = [t.translate(m, scope, context=context)
+                for m in op.metrics]
+        return src_table.agg(*aggs)
+
+
+@compiles(ops.Contains)
+def compile_contains(t, expr, scope, **kwargs):
+    op = expr.op()
+    col = t.translate(op.value, scope)
+    return col.isin(t.translate(op.options, scope))
+
+
+def compile_aggregator(t, expr, scope, fn, context=None, **kwargs):
+    op = expr.op()
+    src_col = t.translate(op.arg, scope)
+
+    if getattr(op, "where", None) is not None:
+        condition = t.translate(op.where, scope)
+        src_col = F.when(condition, src_col)
+
+    col = fn(src_col)
+    if context:
+        return col
+    else:
+        return t.translate(expr.op().arg.op().table, scope).select(col)
```

Comment:
Hm, what are you trying to pull out here with this chain of `op` accesses?

Reply:
Hmm this is a bit convoluted:

The expression is sth like "some_col.max()"

What I need to translate that to is "df.select(max(some_col))"

The `expr.op().arg.op().table` thing is too get to the table of the column. I will add a comment